NE-1366: Revisions for set-delete-http-headers EP (#1446)
openshift/enhancements: enhancements/ingress/set-delete-http-headers.md

#### X-Forwarded-For
`x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
but have actual use-cases (which aren't covered by the existing option). Therefore, they will be allowed, with the following note in the product documentation:

> Note: If in the IngressController, action `Set` is set using API(`spec.httpHeaders.actions`) and if action `Replace` is set using the API mentioned above i.e spec.httpHeaders.forwardedHeaderPolicy then the value will be set
> using this API(spec.httpHeaders.actions).
> In short, the value set by `spec.httpHeaders.actions` will override the value set by `spec.httpHeaders.actions`.

Reviewer: ```suggestion
Note: In the IngressController, if both the action `Set` is configured using `spec.httpHeaders.actions` and the action `Replace` is configured using `spec.httpHeaders.forwardedHeaderPolicy`, then the resulting configured value is from `spec.httpHeaders.actions`.  In short, the value set by `spec.httpHeaders.actions` will override the value set by `spec.httpHeaders.forwardedHeaderPolicy`.
```

Author: ok

Note: In the IngressController, if both the action `Set` is configured using `spec.httpHeaders.actions` and the 
action `Replace` is configured using `spec.httpHeaders.forwardedHeaderPolicy`, then the resulting configured value
is from `spec.httpHeaders.actions`.  
In short, the value set by `spec.httpHeaders.actions` will override the value set by `spec.httpHeaders.forwardedHeaderPolicy`.

In the case of HTTP request headers, the actions specified in `spec.httpHeaders.actions` on the Route will be executed after
the actions specified in the IngressController's `spec.httpHeaders.actions` field.

So, the custom HTTP request value of `x-forwarded-for` header set using `spec.httpHeaders.actions[*].request` of the Route will take over the value defined in annotation `haproxy.router.openshift.io/set-forwarded-headers` of Route, `spec.httpHeaders.forwardedHeaderPolicy` of Ingress Controller
or `spec.httpHeaders.actions[*].request` of Ingress Controller as in case of HTTP request headers, HAProxy overrides the value of the frontend of same header name set by IngressController with the

> value which was present in the backend section.

Reviewer: This sentence needs to be rewrittten.  Maybe make two sentences.  I can't make any sense of this.

Author: I have removed this as now we have a chart.

#### X-Forwarded-For
`x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
but have actual use-cases (which aren't covered by the existing option). Therefore, they will be allowed, with the following note in the product documentation:

Note: In the IngressController, if both the action `Set` is configured using `spec.httpHeaders.actions` and the 
action `Replace` is configured using `spec.httpHeaders.forwardedHeaderPolicy`, then the resulting configured value
is from `spec.httpHeaders.actions`.  

> In short, the value set by `spec.httpHeaders.actions` will override the value set by `spec.httpHeaders.forwardedHeaderPolicy`.

Reviewer: This is different than what is listed in the chart for 

```
spec.httpHeaders.forwardedHeaderPolicy,
spec.httpHeaders.actions in IngressController

```
In the chart, it says the result is:

```
replace policy: 
value in spec.httpHeaders.actions will be replaced by client IP.

```
I think we need to be very careful about what we say in the official Note.

In the case of HTTP request headers, the actions specified in `spec.httpHeaders.actions` on the Route will be executed after
the actions specified in the IngressController's `spec.httpHeaders.actions` field.

> This enhancement provides a method for a user to  delete the `x-forwarded-for` header, which the Ingress Controller `spec.httpHeaders.forwardedHeaderPolicy` does not provide.

Reviewer: Please use `x-forwarded-for` instead of `X-Forwarded-For` below.  Or change them all to one capitalization format.

> This enhancement provides a method for a user to  delete the `x-forwarded-for` header, which the Ingress Controller `spec.httpHeaders.forwardedHeaderPolicy` does not provide.

Reviewer: I would add this to the Note as well (rewording it to remove the reference to an enhancement).

> The `x-forwarded-for` header can be changed in three different places in the IngressControllerSpec and Route annotation: `spec.httpHeaders.actions`, `spec.httpHeaders.forwardedHeaderPolicy`, route annotation `haproxy.router.openshift.io/set-forwarded-headers`

Reviewer: ```suggestion
The `x-forwarded-for` header can be changed in four different places. In the IngressControllerSpec `spec.httpHeaders.actions` and `spec.httpHeaders.forwardedHeaderPolicy`, the Route spec `spec.httpHeaders.actions`,  and the Route annotation `haproxy.router.openshift.io/set-forwarded-headers`
```

> The following chart describes what values will be over-ridden when either `spec.httpHeaders.forwardedHeaderPolicy`, route annotation `haproxy.router.openshift.io/set-forwarded-headers` 
> or both are set along with `spec.httpHeaders.actions`.

Reviewer: ```suggestion
The following chart describes what fields/annotation will set the `x-forwarded-for` header in an environment of multiple combinations.
```

A=haproxy.router.openshift.io/set-forwarded-headers

B=spec.httpHeaders.forwardedHeaderPolicy

C=spec.httpHeaders.actions in IngressController

> D=spec.httpHeaders.actions in Route

Reviewer: Remove.

> | Simplified combination | When this comb. is present                                                                                                                                                  | This value is used                                                                                                               |

Reviewer: Please remove the Simplied combination column, and the legend above.  Also, change:
```suggestion
| When these fields/annotation are set with a value |                                                                                                                                                | This value is used                                                                                                               |
```

| ACD                    | haproxy.router.openshift.io/set-forwarded-headers,spec.httpHeaders.actions in IngressController,spec.httpHeaders.actions in Route                                           | spec.httpHeaders.actions in Route.                                                                                               |
| BCD                    | 	spec.httpHeaders.forwardedHeaderPolicy, spec.httpHeaders.actions in IngressController, spec.httpHeaders.actions in Route                                                   | spec.httpHeaders.actions in Route.                                                                                               |
| AB                     | haproxy.router.openshift.io/set-forwarded-headers, spec.httpHeaders.forwardedHeaderPolicy                                                                                   | haproxy.router.openshift.io/set-forwarded-headers will set the policy for handling X-Forwarded-For headers.                      |

> | AC                     | haproxy.router.openshift.io/set-forwarded-headers, spec.httpHeaders.actions in IngressController                                                                            | spec.httpHeaders.actions will set X-Forwarded-For.                                                                               |

Reviewer: it doesn't make sense that if you set

> haproxy.router.openshift.io/set-forwarded-headers, spec.httpHeaders.forwardedHeaderPolicy, 
> spec.httpHeaders.actions in IngressController
> 

the value is taken from `spec.httpHeaders.actions`, but if you set only
```
spec.httpHeaders.forwardedHeaderPolicy, 
spec.httpHeaders.actions in IngressController
```
the value is sometimes affected by the the policy that is used.  Why does it matter for one case and not the other?  

Same question when using 
```
spec.httpHeaders.forwardedHeaderPolicy, 
spec.httpHeaders.actions in Route
```

Reviewer: @miheer The chart was an idea for expressing the override behavior to users, because  it still isn't clear.  If you have a clear way of expressing the override behavior, let's get rid of the chart, replace it with a clear explanation, and perhaps then we can close this EP.

|                        |                                                                                                                                                                             | never policy:  spec.httpHeaders.actions will set the X-Forwarded-For header.                                                     |
|                        |                                                                                                                                                                             | if-none policy:spec.httpHeaders.actions will set X-Forwarded-For header.                                                         |
| BD	                    | spec.httpHeaders.forwardedHeaderPolicy, spec.httpHeaders.actions in Route                                                                                                   | append policy: spec.httpHeaders.actions will set X-Forwarded-For and one more X-Forwarded-For header will be set with client IP. |

> |                        |                                                                                                                                                                             | replace policy: spec.httpHeaders.actions will set X-Forwaders-For header.                                                        |

Reviewer: ```suggestion
|                        |                                                                                                                                                                             | replace policy: spec.httpHeaders.actions will set x-forwarded-for header                                                       |
```

but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,

> and possibly a warning on admission.

Author: added a note

> and possibly a warning on admission.

Reviewer: I made a suggestion.

In the case of HTTP request headers, the actions specified in `spec.httpHeaders.actions` on the Route will be executed after

> the actions specified in the IngressController's `spec.httpHeaders.actions` field.

Reviewer: I think you need to mention the difference between request and response overrides here, instead of in the questions section.

Author: i did not get this. can you please explain more ?

Reviewer: Is it true that this only matters for request headers, or is there also some special behavior to describe for response headers?   If needed, I think you should point to the  details that are listed in the section "How do Route headers interact with IngressController headers of the same value".  But I wish there was a brief way to state that here.  But maybe this is not needed for x-forwarded-headers?

Author: X-Forwarded headers are request headers so we don't need to describe about response headers from what I understand.

The custom HTTP response value of `x-forwarded` header set using `spec.httpHeaders.actions` of the IngressController will take over the value defined in annotation `haproxy.router.openshift.io/set-forwarded-headers` of Route, `spec.httpHeaders.forwardedHeaderPolicy` of Ingress Controller
or `spec.httpHeaders.actions` of Route as in case of HTTP response headers, HAProxy overrides the value of the backend of same header name which

> was present in the frontend section.

Reviewer 2: Without this enhancement, does HAProxy ever set `x-forwarded-*` in response headers?  The `x-forwarded-*` headers are request headers; given that, none of what you wrote makes any sense.

Author: ok. i tried to update this.

> value which was present in the backend section.

Reviewer: @miheer I think you missed this one, I don't see any changes.

> value which was present in the backend section.

Author: @candita I have created a chart for this.

> Another use case is deleting the header which forwardedHeaderPolicy does not provide but this EP does which is the reason why we will allow users to set/delete

Author: ok

but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
and possibly a warning on admission.

> The custom HTTP response value of `x-forwarded` header set using `spec.httpHeaders.actions` of the IngressController will take over the value defined in annotation `haproxy.router.openshift.io/set-forwarded-headers` of Route, `spec.httpHeaders.forwardedHeaderPolicy` of Ingress Controller

Reviewer: I would add something more explicit, like "In contrast to request headers, in which the IngressController values override the Route `spec.httpHeaders.actions` values, for response headers the Route values override the IngressController values".

Author: Ok added this line above the added line in the diff

Author: i have added a sequence of execution.

or `spec.httpHeaders.actions[*].request` of Ingress Controller as in case of HTTP request headers, HAProxy overrides the value of the frontend of same header name set by IngressController with the
value which was present in the backend section.

This enhancement provides a method for a user to  delete the `x-forwarded-for` header, which the Ingress Controller `spec.httpHeaders.forwardedHeaderPolicy` does not provide.

> | values overriden by spec.httpHeaders.action &rarr;<br/>values set along with spec.httpHeaders.action &darr;                                                                      | haproxy.router.openshift.io/set-forwarded-headers | spec.httpHeaders.forwardedHeaderPolicy |

Reviewer: I don't know why you added this chart, I don't find any place where someone asked for it.  I also don't know what this chart means. It says "yes" in four places, but I don't know what question that answers.  😵‍💫   If you want to keep it, please add an introduction to what it means and clarify what each table column means.  I think it's better to delete it to move this along.

Author: Hi @candita ! You had asked for it. I will try to find out that comment. I am OK to remove this. I am not sure what is confusing you. I will add an introduction.

Author: https://github.com/openshift/enhancements/pull/1446#discussion_r1313260984

Reviewer: This is the kind of chart I was talking about.  Can you use this and fill out the third column?  I still don't know the answer.

| When this combination of values is present                                                                             | This is the value that is used                      |
|------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------|
| haproxy.router.openshift.io/set-forwarded-headers, spec.httpHeaders.forwardedHeaderPolicy, and spec.httpHeaders.action | spec.httpHeaders.action?                            |
| haproxy.router.openshift.io/set-forwarded-headers and spec.httpHeaders.forwardedHeaderPolicy                           | haproxy.router.openshift.io/set-forwarded-headers ? |
| haproxy.router.openshift.io/set-forwarded-headers and spec.httpHeaders.action                                          | ?                                                   |
| spec.httpHeaders.forwardedHeaderPolicy and spec.httpHeaders.action                                                     | spec.httpHeaders.action                             |

Please note that a custom `host` header may be specified via RouteSpec. Please refer to `Open Questions` for more details.

### Cookie and Set-Cookie

> There are three reasons we blocked `Cookie` and `Set-Cookie`:

Reviewer: ```suggestion
We do not allow setting `Cookie` or `Set-Cookie` headers, for the following three reasons:
```

Author: ok

Author: done

1. There were no use cases mentioned for these headers in [RFE](https://issues.redhat.com/browse/RFE-464).

> 2. The cookies that HAProxy sets are used for session tracking, to map a client connection to a particular 

Reviewer: # 2 adds more meaning to the decision than # 1. I would swap the ordering.

Author: done

    backend server.

   The cookie name is an MD5 hash value derived from the backend's name (similarly, the cookie value is a hashed value
   derived from the backend server's endpoint). 
   
   So the cookie names that we would want to forbid are all the possible values
   of the MD5 hashing algorithm, but it's extremely unlikely that someone would accidentally choose the same cookie name

>    that HAProxy used. 

Reviewer: These lines don't add any meaning to the discussion.

Author: but we need to explain what the interference may be right ?

Reviewer: Your comment makes it look like the whole thing is "extremely unlikely" to happen, so I don't think this is the interference we're concerned with.

Author: yes removed

>    However, if there is an interference caused by setting these then either the HAProxy's session affinity might not work correctly,
>    or HAProxy might overwrite the cookie.

Reviewer: ```suggestion
   We don't want to risk interfering with HAProxy's session affinity, nor restrict HAProxy's ownership of a cookie.
```

Author: or HAProxy might overwrite the cookie. sound more better than HAProxy's ownership of a cookie.

Reviewer: What do you suggest?  I wanted to clarify what an "interference" is, as this is not a generally used tech term.

Author: We don't want to risk interfering with HAProxy's session affinity, nor restrict HAProxy's ownership of a cookie. -- added this

> 3. It was thought that it would be better to have a structured API for it.

Reviewer: ```suggestion
3. A better solution for cookie headers is a new structured API feature.
```

Author: done

Please note that a custom `host` header may be specified via RouteSpec. Please refer to `Open Questions` for more details.

> ### Cookie and Set-Cookie

Author: ok

> We do not allow setting `Cookie` or `Set-Cookie` headers, for the following three reasons:

Author: ok

> We do not allow setting `Cookie` or `Set-Cookie` headers, for the following three reasons:

Reviewer: @miheer I think you missed this. I don't see using route annotations as a  fourth reason.

> We do not allow setting `Cookie` or `Set-Cookie` headers, for the following three reasons:

Author: But having an existing annotation is not the reason to deny Cookie and Set-Cookie.

1. The cookies that HAProxy sets are used for session tracking, to map a client connection to a particular 
   backend server.
   We don't want to risk interfering with HAProxy's session affinity, nor restrict HAProxy's ownership of a cookie.

2. There were no use cases mentioned for these headers in [RFE](https://issues.redhat.com/browse/RFE-464).

3. A better solution for cookie headers is a new structured API feature.
  

> The following table summarizes which headers are allowed to be customized via the Route API or IngressController API:

Author: ok

| `strict-transport-security` | No                                                              | No                                                  | `haproxy.router.openshift.io/hsts_header` Route annotation                                                                                   |
| `x-forwarded-for`           | Yes                                                             | Yes                                                 | `haproxy.router.openshift.io/set-forwarded-headers` Route annotation or IngressController `spec.httpHeaders.forwardedHeaderPolicy` API field |
| `x-ssl`                     | Yes                                                             | Yes                                                 | No                                                                                                                                           |        

> | `cookie` or `set-cookie`      | No                                                              | No                                                  | No                                                                                                                                           |        

Reviewer 2: We do have the `haproxy.router.openshift.io/disable_cookies` and `router.openshift.io/cookie_name` route annotations that you could list in this table.

Author: ok

Author: done

  must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
  IngressController.

#### How do Route headers interact with IngressController headers of the same value?
  Answer:  It depends on if they are response headers or request headers.
  

>  In case of HTTP response headers, the headers set in spec.httpHeaders.actions of IngressController will override
>  the headers with same name in Route's spec.httpHeaders.actions or Route annotation.

Reviewer 2: It is more precise to describe the order in which actions are executed.  Also, change "In case" to "In the case".  
```suggestion
 In case of HTTP response headers, the actions specified in `spec.httpHeaders.actions` on the IngressController will be executed after
 the actions specified in the Route's `spec.httpHeaders.actions` field.
```
Please remember to use backticks when referencing API fields.

Reviewer: I agree - please describe the order in which the actions are executed.

Author: yes

Author: done

>  In case of HTTP request headers, the headers set in Route using spec.httpHeaders.actions or annotation will override the 
>  headers with the same name present in the IngressController's spec.httpHeaders.actions.

Reviewer 2: ```suggestion
 In the case of HTTP request headers, the actions specified in `spec.httpHeaders.actions` on the Route will be executed after
 the actions specified in the IngressController's `spec.httpHeaders.actions` field.
```

Author: done

 This is because of the way HAProxy is designed to work. We cannot change the processing order; we're constrained by HAProxy's design: 
 the frontend receives the request and dispatches it to the backend, and then the frontend gets the 
 response from the backend.

> 
>  In case of HTTP response, HAProxy overrides the header value of the same header name in the backend with
>  the value of the header defined in the frontend. So, [the backend rules are applied first, followed by the frontend's rules.](https://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response)

Reviewer 2: The current phrasing seems confusingly repetitive.
```suggestion
 This means that for requests, the frontend's rules are applied first, followed by the backend's,
 and conversely, for responses, [the backend's rules are applied first, followed by the frontend's](https://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response).
```

Author: ok

Author: done

 The `http-response set or delete` stanzas are applied and interpreted in this way by the HAProxy.

> For example: A cluster admin wants to `DENY` response header `X-Frame-Options`. However, a route owner sets  `X-Frame-Options` to `'SAMEORIGIN'` then the haproxy.config will look like the following where

Reviewer 2: The current phrasing seems to suggest that the cluster-admin can deny the router owner from setting a header.  
```suggestion
For example: A cluster admin sets  a response header with the name `X-Frame-Options` and the value `DENY` using the IngressController API. However, a route owner sets  the `X-Frame-Options` header with the value `SAMEORIGIN`.  In this scenario, `haproxy.config` will look like the following where
```

Author: done

the response header called `X-Frame-Options` with value 'DENY' set in frontend will override the value `'SAMEORIGIN'` of `X-Frame-Options`
which was set in the backend section via Route's spec.httpHeaders.actions.
```conf
frontend public
  http-response set-header X-Frame-Options 'DENY' 

frontend fe_sni
  http-response set-header X-Frame-Options 'DENY' 

frontend fe_no_sni
  http-response set-header X-Frame-Options 'DENY' 

backend be_secure:openshift-monitoring:alertmanager-main
  http-response set-header X-Frame-Options 'SAMEORIGIN'

> ```

Reviewer 2: ```mermaid
sequenceDiagram
    client->>frontend: request (IngressController spec.httpHeaders.actions[].request[] applied)
    frontend->>backend: request (Route spec.httpHeaders.actions[].request applied)
    backend->>application: request (application receives request headers)
    application->>backend: response (application sets response headers)
    backend->>frontend: response (Route spec.httpHeaders.actions[].response applied)
    frontend->>client: response (IngressController IngressController spec.httpHeaders.actions[].response applied)
```

Reviewer: Maybe add Route annotation processing too.

Reviewer: I learned you can add a line break using <br> in the text, making it take up less space horizontally.

sequenceDiagram
    client->>frontend: request (IngressController <br> spec.httpHeaders.actions[].request[] applied)
    frontend->>backend: request (Route <br> spec.httpHeaders.actions[].request applied)
    backend->>application: request (application <br> receives request headers)
    application->>backend: response (application <br> sets response headers)
    backend->>frontend: response (Route <br> spec.httpHeaders.actions[].response applied)
    frontend->>client: response (IngressController <br> IngressController spec.httpHeaders.actions[].response applied)

Author: sequenceDiagram
    client->>frontend: request (IngressController spec.httpHeaders.actions[].request[] applied)
    frontend->>backend: request (Route spec.httpHeaders.actions[].request applied. If in presence of route annotations for the same header, spec.httpHeaders.actions[].request will be applied after it.)
    backend->>application: request (application receives request headers)
    application->>backend: response (application sets response headers)
    backend->>frontend: response (Route spec.httpHeaders.actions[].response applied. If in presence of route annotations for the same header, spec.httpHeaders.actions[].request will be applied after it.)
    frontend->>client: response (IngressController IngressController spec.httpHeaders.actions[].response applied)

Author: ```mermaid
sequenceDiagram
    client->>frontend: request (IngressController spec.httpHeaders.actions[].request[] applied)
    frontend->>backend: request (Route spec.httpHeaders.actions[].request applied. If in presence of route annotations for the same header, spec.httpHeaders.actions[].request will be applied after it.)
    backend->>application: request (application receives request headers)
    application->>backend: response (application sets response headers)
    backend->>frontend: response (Route spec.httpHeaders.actions[].response applied. If in presence of route annotations for the same header, spec.httpHeaders.actions[].request will be applied after it.)
    frontend->>client: response (IngressController IngressController spec.httpHeaders.actions[].response applied)
```

 However, in case of the HTTP request, HAProxy overrides the values of the header value of the same header name present in the frontend with the value of the
 header defined in backend. [The rules are evaluated in their declaration order when they are met in a frontend,

>  listen or backend section.](https://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-request)

Reviewer 2: How is this statement relevant here?

Author: I just tried to give more information. However I removed this and added a link of http-request -
 This means that for requests, [the frontend's rules are applied first, followed by the backend's,
and conversely](https://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-request),

  must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
  IngressController.

#### How do Route headers interact with IngressController headers of the same value?
  Answer:  It depends on if they are response headers or request headers.

In case of HTTP response headers, the actions specified in `spec.httpHeaders.actions` on the IngressController will be executed after

> the actions specified in the Route's `spec.httpHeaders.actions` field.

Author: ok

In the case of HTTP request headers, the actions specified in `spec.httpHeaders.actions[*].request` on the Route will be executed after
the actions specified in the IngressController's `spec.httpHeaders.actions[*].request` field.

 This is because of the way HAProxy is designed to work. We cannot change the processing order; we're constrained by HAProxy's design: 
 the frontend receives the request and dispatches it to the backend, and then the frontend gets the 
 response from the backend.

This means that for requests, [the frontend's rules are applied first, followed by the backend's,
and conversely](https://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-request), for responses, [the backend's rules are applied first, followed by the frontend's](https://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response).

>  The `http-response set or delete` stanzas are applied and interpreted in this way by the HAProxy.

Reviewer: Why does this only mention http-response?  I think this sentence can be deleted?

Author: The `http-response/http-request set or delete` stanzas are applied and interpreted in this way by the HAProxy.
what about this ?

> For example: A cluster admin sets a response header with the name `X-Frame-Options` and the value `DENY` using the IngressController API. However, a route owner sets  the `X-Frame-Options` header with the value `SAMEORIGIN`.
> 
> In this scenario, `haproxy.config` will look like the following where

Reviewer: I think you should add a line break before "For example" and remove the line break before "In this scenario".

```suggestion

For example: A cluster admin sets a response header with the name `X-Frame-Options` and the value `DENY` using the IngressController API. However, a route owner sets  the `X-Frame-Options` header with the value `SAMEORIGIN`.  In this scenario, `haproxy.config` will look like the following where
```

Author: ok

> the response header called `X-Frame-Options` with value 'DENY' set in frontend will override the value `'SAMEORIGIN'` of `X-Frame-Options`
> which was set in the backend section via Route's spec.httpHeaders.actions.

Reviewer: Seems to mean the way I wrote it below, is that correct?

```suggestion
the response header called `X-Frame-Options` with value 'DENY' applied to the frontend configurations will override the response header `X-Frame-Options` with the value `'SAMEORIGIN'` set in the backend section.
```

Author: ok that looks fine

```conf
frontend public
  http-response set-header X-Frame-Options 'DENY' 

frontend fe_sni
  http-response set-header X-Frame-Options 'DENY' 

frontend fe_no_sni
  http-response set-header X-Frame-Options 'DENY' 

backend be_secure:openshift-monitoring:alertmanager-main
  http-response set-header X-Frame-Options 'SAMEORIGIN'
```

The following sequence diagram represents how the headers get set in sequence and how they could get over-ridden:

> ```mermaid
> sequenceDiagram
>     client->>frontend: request (IngressController spec.httpHeaders.actions[].request[] applied)
>     frontend->>backend: request (Route spec.httpHeaders.actions[].request applied. If in presence of route annotations for the same header, spec.httpHeaders.actions[].request will be applied after it.)
>     backend->>application: request (application receives request headers)
>     application->>backend: response (application sets response headers)
>     backend->>frontend: response (Route spec.httpHeaders.actions[].response applied. If in presence of route annotations for the same header, spec.httpHeaders.actions[].request will be applied after it.)
>     frontend->>client: response (IngressController IngressController spec.httpHeaders.actions[].response applied)
> ```

Reviewer: Add the < b r > as mentioned so the diagram fits and can be read.  Also, I think you typed **request** where it should be **response** in one section.
```suggestion
```mermaid
sequenceDiagram
    client->>frontend: request - IngressController <br> spec.httpHeaders.actions[].request[] applied
    frontend->>backend: request - Route spec.httpHeaders.actions[].request applied. <br> If in presence of route annotations for the same header,<br>  spec.httpHeaders.actions[].request will be applied after it.
    backend->>application: request <br> (application receives request headers)
    application->>backend: response <br> (application sets response headers)
    backend->>frontend: response - Route spec.httpHeaders.actions[].response applied. <br> If in presence of route annotations for the same header, <br> spec.httpHeaders.actions[].response will be applied after it.
    frontend->>client: response - IngressController <br> IngressController spec.httpHeaders.actions[].response applied
```
```

Author: ok

  IngressController.

#### Why does a custom header value defined in RouteSpec override IngressControllerSpec value ?
  Answer: It's quite common to have a global setting with an escape hatch locally to override the global setting.
  The purpose of this EP is to define custom headers. So, there are two reasons we need such override:
  1. A route owner shall have the choice to define what headers he wants as per his application requirements(as they might be varying) 
  than the ones admin has set cluster-wide.
  2. The IngressController and Route might be owned by the same person who wants to specify some global configuration

>   and then override that global configuration on a small subset of routes. 

Reviewer: I would keep this section in here, not move it to the Alternatives section.

Author: i think this comes under upside

Reviewer: I don't understand?

Author: i think we will need to remove this as we no-longer use this over-ride behavior where route over-rides ingresscontroller in all cases.

>   Answer: In case of HTTP response headers, the headers set in spec.httpHeaders.actions of IngressController will override

Reviewer: For the title of this section, I would clarify it by stating: "How do Route headers interact with IngressController headers of the same value?"

Author: done

>  This is because of the way Haproxy is designed to work. We cannot change the processing order; we're constrained by HAProxy's design: 

Reviewer: nit: use the same capitalization for HAProxy throughout.
```suggestion
 This is because of the way HAProxy is designed to work. We cannot change the processing order; we're constrained by HAProxy's design: 
```

Author: done

 the frontend receives the request and dispatches it to the backend, and then the frontend gets the 
 response from the backend.

 In case of HTTP response, Haproxy overrides the header value of the same header name in the backend with

>  the value of the header defined in the frontend. So, [the backend rules are applied first, followed by the frontend's rules.](https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#4.2-http-response)

Reviewer: nit: update the url to point to 2.6 documentation, not 1.7

Author: done

 The `http-response set or delete` stanzas are applied and interpreted in this way by the haproxy.

 However, in case of the HTTP request, haproxy overrides the values of the header value of the same header name present in the frontend with the value of the
 header defined in backend. [The rules are evaluated in their declaration order when they are met in a frontend,
 listen or backend section.](https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#4.2-http-request)

>  So, the `http-request set or delete` stanzas are applied and interpreted this way by the haproxy.

Reviewer: Add an example of how to explicitly set the headers for a desired result, both for request and response.  I assume this means that they have to set request headers on Route and corresponding response headers on IC if they want predictability in the entire transaction.

Author: they have to set request headers on Route and corresponding response headers ---
You can't set response header in http-request stanzas and vice-versa. haproxy does not allow that.
For example you can set response header X-Frame-Options in request list ad it will get added in http-request stanza which is not allowed by haproxy.

Reviewer: Is that what you were testing recently?

Author: Yes.....I take my words back.

#### Does a custom header value defined in RouteSpec override the route annotation ?

>   Answer: Yes. A value defined in RouteSpec overrides the route annotation.

Reviewer: i would add: except for response headers, where the IngressControllerSpec overrides both.

Author: ok

Author: done

    value was set.
- If the API fields are not getting set via IngressController Spec or Route Spec i.e. if `oc patch` is failing to edit values of the Ingress Controller or Route CR 
  then check the openshift-apiserver logs. You can even run `oc explain route.spec.httpHeaders` or `oc explain ingresscontroller.spec.httpHeaders` to understand how to set the `httpHeaders` field.

> - Check [access logs](https://docs.openshift.com/container-platform/4.12/rest_api/operator_apis/ingresscontroller-operator-openshift-io-v1.html#spec-logging-access) of the HAProxy if you want to check what HTTP Response or Request was set for the frontend and backend.

Reviewer: ```suggestion
- Check the [access logs](https://docs.openshift.com/container-platform/4.12/rest_api/operator_apis/ingresscontroller-operator-openshift-io-v1.html#spec-logging-access) of the HAProxy if you want to check what HTTP Response or Request was set for the frontend and backend.
```

Author: ok done

  2. Wrap the entire string in single-quotes.
The HAProxy documentation refers to this as _strong quoting_.  Using strong quoting allows us to keep the quoting logic simple and `haproxy.config` more human-readable compared to the alternatives.

### Proposed Alternative Solution for overriding the HTTP response headers defined in the IngressController by the Route
As per HAProxy's design: the frontend receives the request and dispatches it to the backend, and then the frontend
gets the response from the backend.

So, the override behavior happens the way mentioned in the question `Why does a custom header value defined in RouteSpec override IngressControllerSpec value ?`

> under section `Open Questions` where HTTP Response headers defined in the IngressController are overridden by headers 

Reviewer: ```suggestion
under section `Open Questions` where HTTP response headers defined in the IngressController are overridden by headers 
```

Author: done

defined in the Route.

However, if we want the same behavior of override for HTTP response headers as that of HTTP request headers then we will have
to inject the set-header from the IngressController config into every backend and not the frontend.

> The downside is that it would grow the generated haproxy.config significantly. 

Reviewer: Can you be more explicit?  Would it grow it by one line per backend or would it grow it by one line per affected backend?

Author: check line 1167 below

Author: Yes it is there in ### Proposed Alternative Solution for overriding the HTTP response headers defined in the IngressController by the Route

The upside is:

It would give us better control over the ordering.

> It would eliminate the issue with the host header. So, we could allow it to be set via IngressController.

Reviewer: Can you explain a little why it eliminates the issue?
```suggestion
Because it prevents _____, we could allow the host header to be set via IngressController.
```

Author: updated.

The reason for this proposal was that it's quite common to have a global setting with an escape hatch locally to 
override the global setting.
The purpose of this EP is to define custom headers. So, there are two reasons we need such override:
1. A route owner shall have the choice to define what headers he wants as per his application requirements(as they might be varying)
   than the ones admin has set cluster-wide.
2. The IngressController and Route might be owned by the same person who wants to specify some global configuration
   and then override that global configuration on a small subset of routes.

> 

Reviewer: This doesn't belong here, it doesn't add to the meaning.

Author: it does not exist now.

As per HAProxy's design: the frontend receives the request and dispatches it to the backend, and then the frontend
gets the response from the backend.

So, the override behavior happens the way mentioned in the question `Why does a custom header value defined in RouteSpec override IngressControllerSpec value ?`
under section `Open Questions` where HTTP request headers defined in the IngressController are overridden by headers 
defined in the Route.

However, if we want the same behavior of override for HTTP response headers as that of HTTP request headers then we will have
to inject the set-header from the IngressController config into every backend and not the frontend.

The downside is that it would grow the generated haproxy.config significantly. 
For Example: Suppose an ingress controller called custom-header handles 100 routes.
If it sets HTTP response X-Frame-Options DENY via spec.httpHeader.actions, 
then for all those 100 routes `http-response set-header X-Frame-Options DENY` will be set i.e. 100 stanzas as it will be one stanza per backend related to a route. 

> There might be a case, where admin wants to set a response header for all connections to the IngressController as per company policy which he won't be able to do so
> using this approach.
> For Example: A cluster admin wants to deny X-Frame-Options for all routes the ingress controller is handling.

Reviewer 2: This seems like a critical advantage to the current design.  
```suggestion
There might be a case, where admin wants to set a response header for all connections to the IngressController as per company policy which he won't be able to do so
using this approach.
For Example: A cluster admin wants to set the  `X-Frame-Options` header with the value `DENY` for all routes the ingress controller is handling, overwriting any existing `X-Frame-Options` header that may have been set by the application or Route configuration.
```

Author: done

> The upside is:
> 
> It would give us better control over the ordering.

Reviewer 2: Upside of what?  What would?  Give whom control?  Don't force the reader to guess.  
```suggestion
The upside of putting all the header configuration on the backends is that
it would give OpenShift router full control over the ordering of header actions.
```

Author: sorry. thanks for the improvement!

Author: done

> Because it prevents an issue as mentioned in the section `Host` under `Risks and Mitigations` section, we could allow the host header to be set via IngressController
> as we won't be setting headers in the frontend sections but backend section.
> So, we could allow it to be set via IngressController.

Reviewer 2: You might as well spell out the issue.
```suggestion
In addition, in the case of the `host` header, because the actions would take place after the frontend dispatches the connection to the backend, we could allow the `host` header to be modified via the IngressController API (see `Risks and Mitigations` section).
```

Author: done

The reason for this proposal was that it's quite common to have a global setting with an escape hatch locally to 

> override the global setting.

Reviewer 2: The reason for what proposal?  The EP as a whole, or this specific alternative?  Is "escape hatch" really an appropriate term?

Author: We will need to remove this as our over-ride approach is not longer the same for response headers.

  2. Wrap the entire string in single-quotes.
The HAProxy documentation refers to this as _strong quoting_.  Using strong quoting allows us to keep the quoting logic simple and `haproxy.config` more human-readable compared to the alternatives.

### Proposed Alternative Solution for overriding the HTTP response headers defined in the IngressController by the Route

> As per HAProxy's design: the frontend receives the request and dispatches it to the backend, and then the frontend

Reviewer: ```suggestion
By design, HAProxy's frontend receives the request and dispatches it to the backend, and then the frontend
```

Author: ok done

gets the response from the backend.

So, the override behavior happens the way mentioned in the question `Why does a custom header value defined in RouteSpec override IngressControllerSpec value ?`
under section `Open Questions` where HTTP request headers defined in the IngressController are overridden by headers 

> defined in the Route.

Reviewer: ```suggestion
defined in the Route.  HTTP response headers defined in the Ingress Controller are **not** overridden by response headers defined in the Route.
```

Author: ok done

> However, if we want the same behavior of override for HTTP response headers as that of HTTP request headers then we will have

Reviewer: ```suggestion
We considered implementing the same behavior of override for HTTP response headers as that of HTTP request headers.  However, to do this we would have
```

Author: ok done

to inject the set-header from the IngressController config into every backend and not the frontend.

> The downside is that it would grow the generated haproxy.config significantly. 

Reviewer: ```suggestion
We rejected this because it would grow the generated haproxy.config significantly, compromising the performance of HAProxy. 
```

Author: ok done

For Example: Suppose an ingress controller called custom-header handles 100 routes.

> If it sets HTTP response X-Frame-Options DENY via spec.httpHeader.actions, 

Reviewer: ```suggestion
If it sets HTTP response `X-Frame-Options DENY` via `spec.httpHeader.actions`, 
```

Author: ok done

> then for all those 100 routes `http-response set-header X-Frame-Options DENY` will be set i.e. 100 stanzas as it will be one stanza per backend related to a route. 

Reviewer: ```suggestion
then for each of those 100 routes the command `http-response set-header X-Frame-Options DENY` will be set, resulting in 100 additional stanzas as it will be one stanza per backend related to a route. 
```

Author: ok done

> There might be a case, where admin wants to set a response header for all connections to the IngressController as per company policy which he won't be able to do so
> using this approach.
> For Example: A cluster admin wants to set the  `X-Frame-Options` header with the value `DENY` for all routes the ingress controller is handling, overwriting 
> any existing `X-Frame-Options` header that may have been set by the application or Route configuration.

Reviewer: I think you mean this, as reworded, but I'm not sure how it could be done by the current proposal.  Can it be done with the current proposal, e.g. can you fill in the blank?
```suggestion
There might be a case, where company policy requires the admin to set a response header for all connections to the IngressController.  
For example, a cluster admin wants to set the  `X-Frame-Options` header with the value `DENY` for all routes the ingress controller is handling, overwriting 
any existing `X-Frame-Options` header that may have been set by the application or Route configuration.  This wouldn't be possible with this alternative solution.  However, it can be done by the current proposal by _______.
```

Author: But why do need to mention all the steps like adding headers in IC and route ? We have already mentioned them in this EP.

Reviewer: Then you could just say "as described in the section ____".

My re-write also removed the "he" pronoun, which has generally been removed from most content for at least a decade, if not longer.

Author: ok done

> The upside of putting all the header configuration on the backends is that
> it would give OpenShift router full control over the ordering of header actions.

Reviewer: ```suggestion
The upsides of injecting all the header configuration on the backends is that
it would give OpenShift router full control over the ordering of header actions and present a more consistent learning environment for the admin.
```

Author: ok done

In addition, in the case of the `host` header, because the actions would take place after the frontend dispatches the connection to the backend, 

> we could allow the `host` header to be modified via the IngressController API (see `Risks and Mitigations` section).

Reviewer: There are more pros than cons.  It makes it look strange that we chose not to do that.

Author: We already discussed this on our calls and agreed to the approach we currently got merged. The current design works as per Haproxy design.

  2. Wrap the entire string in single-quotes.
The HAProxy documentation refers to this as _strong quoting_.  Using strong quoting allows us to keep the quoting logic simple and `haproxy.config` more human-readable compared to the alternatives.

### Proposed Alternative Solution for overriding the HTTP response headers defined in the IngressController by the Route
By design, HAProxy's frontend receives the request and dispatches it to the backend, and then the frontend
gets the response from the backend.

> So, the override behavior happens the way mentioned in the question `Why does a custom header value defined in RouteSpec override IngressControllerSpec value ?`

Reviewer: The question title was renamed:
```suggestion
So, the override behavior happens the way mentioned in the question `How do Route headers interact with IngressController headers of the same value?`
```

Author: ok done
